src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheck.java

```diff
@@ -189,33 +189,48 @@ public void setAllowEmptyLoopBody(boolean allowEmptyLoopBody) {
 
     @Override
     public void visitToken(DetailAST ast) {
-        final DetailAST slistAST = ast.findFirstToken(TokenTypes.SLIST);
-        boolean isElseIf = false;
-        if (ast.getType() == TokenTypes.LITERAL_ELSE
-            && ast.findFirstToken(TokenTypes.LITERAL_IF) != null) {
-            isElseIf = true;
+        final int type = ast.getType();
+        final boolean hasNoSlist = ast.findFirstToken(TokenTypes.SLIST) == null;
+        boolean hasViolation = hasNoSlist && !isSkipStatement(ast);
+        switch (type) {
+            case TokenTypes.LITERAL_FOR:
+            case TokenTypes.LITERAL_WHILE:
+                hasViolation = hasViolation && !isEmptyLoopBodyAllowed(ast);
+                break;
+            case TokenTypes.LITERAL_CASE:
+            case TokenTypes.LITERAL_DEFAULT:
+                hasViolation = hasViolation && hasUnbracedStatements(ast);
+                break;
+            case TokenTypes.LITERAL_ELSE:
+                hasViolation = hasViolation && ast.findFirstToken(TokenTypes.LITERAL_IF) == null;
+                break;
+            default:
+                break;
         }
-        final boolean isInAnnotationField = isInAnnotationField(ast);
-        final boolean skipStatement = isSkipStatement(ast);
-        final boolean skipEmptyLoopBody = allowEmptyLoopBody && isEmptyLoopBody(ast);
-
-        if (slistAST == null && !isElseIf && !isInAnnotationField
-                && !skipStatement && !skipEmptyLoopBody) {
+        if (hasViolation) {
             log(ast.getLineNo(), MSG_KEY_NEED_BRACES, ast.getText());
         }
     }
 
     /**
-     * Checks if ast is in an annotation field.
-     * @param ast ast to test.
-     * @return true if current ast is part of an annotation field.
+     * Checks if current loop has empty body and can be skipped by this check.
+     * @param ast for, while statements.
+     * @return true if current loop can be skipped by check.
      */
-    private static boolean isInAnnotationField(DetailAST ast) {
-        boolean isDefaultInAnnotation = false;
-        if (ast.getParent().getType() == TokenTypes.ANNOTATION_FIELD_DEF) {
-            isDefaultInAnnotation = true;
-        }
-        return isDefaultInAnnotation;
+    private boolean isEmptyLoopBodyAllowed(DetailAST ast) {
+        return allowEmptyLoopBody && ast.findFirstToken(TokenTypes.EMPTY_STAT) != null;
+    }
+
+    /**
+     * Checks if switch member (case, default statements) has statements without curly braces.
+     * @param ast case, default statements.
+     * @return true if switch member has unbraced statements, false otherwise.
+     */
+    private static boolean hasUnbracedStatements(DetailAST ast) {
+        final DetailAST nextSibling = ast.getNextSibling();
+        return nextSibling != null
+            && nextSibling.getType() == TokenTypes.SLIST
+            && !"{".equals(nextSibling.getFirstChild().getText());
```

Comment:
This can be
```
        return nextSibling != null
            && nextSibling.getType() == TokenTypes.SLIST
            && nextSibling.getFirstChild().getType() != TokenTypes.SLIST;
```

Reply:
done